pype/mongodb.py

```diff
@@ -0,0 +1,382 @@
+                "ERROR: Couldn't connect to {} in less than {:.3f}ms"
+            ).format(mongo_url, timeout))
+
+        cls.log.info("Connected to {}, delay {:.3f}s".format(
+            mongo_url, time.time() - t1
+        ))
+        return mongo_client
+
+
+class PypeMongoDB:
+    """ Enhanced wrapper around pymongo.database used for all operations
+        based on pymongo.database.
+        Has a functionality for auto_connect, parent mongo client is
+        accessible.
+    """
+    def __init__(self, collection, auto_install=True):
+        self._id = uuid4()
+        self._database = None
+        self.auto_install = auto_install
+        self._collection = collection
+
+        self.log = logging.getLogger(self.__class__.__name__)
+
+    def __getattr__(self, attr_name):
+        """
+            All not explicitly implemented operations on pymongo.database
+            are wrapped in 'auto_reconnect' decorator
+        Args:
+            attr_name (string): attribute from pymongo.database
+
+        Returns:
+            attribute or wrapped callable
+        """
+        print("attr_name:: {}".format(attr_name))
+        self.log.debug("attr_name:: {}".format(attr_name))
+        attr = None
+        if self.is_installed() and self.auto_install:
+            self.install()
+
+        if self.is_installed():
+            attr = getattr(
+                self._database[self._collection],
+                attr_name,
+                None
+            )
+
+        if attr is None:
+            # Reraise attribute error
+            return self.__getattribute__(attr_name)
+
+        # Decorate function
+        if callable(attr):
+            attr = auto_reconnect(attr)
+        return attr
+
+    @property
+    def mongo_client(self):
+        """
+        Returns:
+            pymongo.MongoClient
+        """
+        return PypeMongoConnection.mongo_client()
+
+    @property
+    def id(self):
+        """
+            Identifier of database
+        Returns:
+            (uuid4)
+        """
+        return self._id
+
+    @property
+    def database(self):
+        if not self.is_installed() and self.auto_install:
+            self.install()
+
+        if self.is_installed():
+            return self._database
+
+        raise IOError(
+            "'{}.database' requires to run install() first".format(
+                self.__class__.__name__
+            )
+        )
+
+    def is_installed(self):
+        return PypeMongoConnection.is_installed(self)
+
+    def install(self):
+        """Establish a persistent connection to the database"""
+        if self.is_installed():
+            return
+
+        PypeMongoConnection.install(self)
+
+        self._database = PypeMongoConnection.database()
+
+    def uninstall(self):
+        """Close any connection to the database"""
+        PypeMongoConnection.uninstall(self)
+        self._database = None
+
+    @auto_reconnect
+    def insert_one(self, item, *args, **kwargs):
```

Comment:
`insert_one` and `insert_many` are in avalon only because of document validation on insertion, which does not make sence here. I would remove both as they are handled in `__getattr__` and remove import `from avalon import schema`

Reply:
I agree, that here it is not needed.